src/slices.md

```diff
@@ -0,0 +1,260 @@
+}
+```
+In the above example, we obtain permission to the slice elements on allocation by making them.
+We obtain permission to the slice elements on allocation,
+in the above example by making it.
+``` go
+s1 := []int{0, 1, 1, 2, 3, 5}
+//@ assert forall i int :: {&s1[i]} 0 <= i && i < len(s1) ==> acc(&s1[i])
+//@ assert acc(s1)
+```
+After initialization with a literal, we also gain permission.
+The second assertion is syntactic sugar for the first.
+## Slicing
+A slice can be created from another slice or array by slicing it, in general with an expression of the form `a[i:j:k]`.
+To avoid panics, as part of the contract of the slicing operation, Gobra checks that `0 <= i && i <= j && j <= k && k <= cap(a)`.
+
+We can create two overlapping slices `l` and `r` but run into a permission error:
+``` go
+~func client2() {
+    s := make([]int, 3)
+    l := s[:2]
+    r := s[1:]
+    addToSlice(l, 1)
+    addToSlice(r, 1)
+~}
+```
+``` go
+ERROR Assert might fail. 
+Permission to r might not suffice.
+```
+While we should not be able to assert `acc(l) && acc(r)` (why?),
+to call `addToSlice` we only need permission for either of the slices.
+In order to get permission for the slice `r` we must explicitly assert how the addresses to the elements of `r` relate to those of `s`.
+``` go
+func client2() {
+	s := make([]int, 3)
+	l := s[:2]
+	r := s[1:]
+	//@ assert forall i int :: {&r[i]} 0 <= i && i < len(r) ==> &r[i] == &s[i+1]
+	addToSlice(l, 1)
+	//@ assert s[0] == 1 && s[1] == 1
+	addToSlice(r, 1)
+	//@ assert r[0] == 2 && r[1] == 1
+	//@ assert s[0] == 1 && s[1] == 2 && s[2] == 1
+}
+```
+
+## Copy and Append
+Go includes the `copy` and `append` built-in functions for common slice operations.
+
+We give the contracts for `copy` and `append` for a generic type `T` (Gobra does not yet support generics).
+In Gobra we must pass an additional ghost parameter for the permission amount required to read the elements of `src`.
+This allows the functions generally to be used independent of the exact permission amount.
+``` gobra
+requires p > 0
+requires forall i int :: {&src[i]} 0 <= i && i < len(src) ==> acc(&src[i])
+requires forall i int :: {&xs[i]} 0 <= i && i < len(xs) ==> acc(&xs[i], p)
+ensures len(res) == len(src) + len(xs)
+ensures forall i int :: {&res[i]} 0 <= i && i < len(res) ==> acc(&res[i])
+ensures forall i int :: {&xs[i]} 0 <= i && i < len(xs) ==> acc(&xs[i], p)
+ensures forall i int :: {&res[i]} 0 <= i && i < len(src) ==> res[i] === old(src[i])
+ensures forall i int :: {&res[i]} len(src) <= i && i < len(res) ==> res[i] === xs[i - len(src)]
+func append[T any](ghost p perm, src []T, xs ...T) (res []T)
+```
+
+Note that since `append` is variadic, the permission must be the first argument.
+The permission to `src` is lost as the underlying array could be reallocated if the capacity is not sufficient to append the new elements.
+`s = append(s, 42)` is a common pattern in Go and this way permissions to `s` are preserved.
+
```

Comment:
given that variadic functions are treated in the same way as functions that take a slice as an input, it may be worth having a section on this chapter on variadic functions

Reply:
added as a TODO section for now